### PR TITLE
Add tag support to consultant

### PR DESCRIPTION
**Pull request by rogierslag**
Supersedes #11 

The goal of adding tags to Consultant is the way we can tag instances. Examples are the built time, the git sha, or whether this instance is allowed to receive production traffic. Other infrastructure services down the road can inspect and use these tags to determine where to send traffic.

@Magnetme/monolith ready for review!

### CODE REVIEW

File: src/main/java/me/magnet/consultant/Consultant.java [hunk 1]
```diff
@@ -243,6 +246,13 @@ public Builder identifyAs(String serviceName, String datacenter, String hostname
 			return this;
 		}
 
+		public Builder withTag(String tag) {
+			if ( tag != null) {
```
Comment (TiddoLangerak) [inline src/main/java/me/magnet/consultant/Consultant.java]: Shouldn't you just throw here, or maybe even let an NPE be triggered? If `tag == null` then that's most likely a programming error.

Comment (alexnederlof) [inline src/main/java/me/magnet/consultant/Consultant.java]: Use `Preconditions.checkNotNull(tag, "Tag cannot be null")`

Comment (rogierslag) [inline src/main/java/me/magnet/consultant/Consultant.java]: Same for empty probably

Comment (alexnederlof) [inline src/main/java/me/magnet/consultant/Consultant.java]: Yes, in fact, what's even more correct: `Preconditions.checkArgument(!Strings.isNullOrEmpty(tag), "enter a non-empty tag name")

Comment (rogierslag) [inline src/main/java/me/magnet/consultant/Consultant.java]: All handled in the next commit (with additional tests)

File: src/main/java/me/magnet/consultant/Consultant.java [hunk 2]
```diff
@@ -356,6 +367,9 @@ public Consultant build() {
 			instanceName = Optional.ofNullable(instanceName)
 					.orElse(Optional.ofNullable(fromEnvironment("SERVICE_INSTANCE"))
 							.orElse(UUID.randomUUID().toString()));
+			Arrays.stream(Strings.nullToEmpty(fromEnvironment("SERVICE_TAGS")).split(","))
+					.filter((t) -> t != null && !t.isEmpty())
```
Comment (AlexKolpa) [inline src/main/java/me/magnet/consultant/Consultant.java]: `.filter(tag -> !Strings.isNullOrEmpty(tag))`

File: src/main/java/me/magnet/consultant/Consultant.java [hunk 3]
```diff
@@ -356,6 +367,9 @@ public Consultant build() {
...
+					.forEach((tag) -> tags.add(tag));
```
Comment (AlexKolpa) [inline src/main/java/me/magnet/consultant/Consultant.java]: tags::add

File: src/main/java/me/magnet/consultant/Consultant.java (hunk 2 above)
Comment (rogierslag) [inline src/main/java/me/magnet/consultant/Consultant.java]: Bit of a bummer Java does not allow method reference negations eg `filter(!Strings::isNullOrEmpty)` so we'll do it the older fashioned way

Comment (AlexKolpa) [inline src/main/java/me/magnet/consultant/Consultant.java]: when I have a lot of streams that need negation I usually just copy-paste these lines:
```
public static Function<T, Boolean> not(Function<T, Boolean> function) {
   return (input) -> !function.apply(input);
}
````

Quick fix, and let's me do `filter(not(Strings::isNullOrEmpty))`. But I'd say that might be a bit much for this PR :wink:

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java [hunk 4]
```diff
@@ -123,6 +138,7 @@ public String toString() {
 		getDatacenter().ifPresent(dc -> descriptors.add("dc=" + dc));
 		getHostName().ifPresent(host -> descriptors.add("host=" + host));
 		getInstance().ifPresent(instance -> descriptors.add("instance=" + instance));
+		getTags().stream().forEach((tag) -> descriptors.add("tag=" + tag));
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: If a service has multiple tags, this becomes something like `oauth[dc=eu-central,host=docker0,instance=<uuid>,tag=192,tag=staging,tag=testing]` etc. Wouldn't it be far nicer (and easier to parse) something like: `oauth[dc=eu-central,host=docker0,instance=<uuid>,tags=192,staging,testing]`.

**Comment by alexnederlof** [comment]
What @michaeldejong said.

👍 besides that

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java (hunk 4 above)
Comment (rogierslag) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: In that case (which I'm fine with) I'm opting for a result similar to `oauth[dc=eu-central,host=docker0,instance=<uuid>,tags=[192,staging,testing]]`. This makes it clearer it represents a set. Additionally this is a lot clearer when dealing with an empty set of tags.

**Comment by rogierslag** [comment]
All feedback handled and rebased!

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java (hunk 4 above)
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: Well ideally when no tags are set it shouldn't be shown at all. The same principle applies to the other variables in the `ServiceIdentifier`.

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java [hunk 5]
```diff
@@ -47,6 +56,10 @@ public String getServiceName() {
 		return instance;
 	}
 
+	public Set<String> getTags() {
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: Return an `ImmutableSet` instead. This allows others to manipulate the contents of the `Set` post-construction.

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java [hunk 6]
```diff
@@ -123,6 +138,7 @@ public String toString() {
 		getDatacenter().ifPresent(dc -> descriptors.add("dc=" + dc));
 		getHostName().ifPresent(host -> descriptors.add("host=" + host));
 		getInstance().ifPresent(instance -> descriptors.add("instance=" + instance));
+		descriptors.add("tags=[" + getTags().stream().collect(Collectors.joining(",")) + "]");
```
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: Put an if statement around this, and remove the brackets. The brackets are likely to mess up other things like setting keys, and you should only show this if you actually defined tags.

**Comment by michaeldejong** [comment]
Very close. Just some minor dotting the i's, and then we're all good 👍

**Comment by michaeldejong** [comment]
If you can handle this today, I can put out a new release asap. 😉

File: src/main/java/me/magnet/consultant/ServiceIdentifier.java (hunk 6 above)
Comment (michaeldejong) [inline src/main/java/me/magnet/consultant/ServiceIdentifier.java]: Although without brackets it might be a bit hard to parse. Has this broken any tests related to reading config?

**Comment by michaeldejong** [comment]
Reopen when needed.
